### PR TITLE
Initial work to get 'dependent type finding' moved out of process.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
This is a followup to https://github.com/dotnet/roslyn/pull/43018

Primarily this entailed ensuring that all entrypoints take a SymbolAndProjectId so that we can serialize/deserialize that to OOP to actually do the work.   This hit progression, which involved a bunch of viral changes.

### CODE REVIEW

File: src/Features/Core/Portable/GenerateMember/GenerateParameterizedMember/TypeParameterSubstitution.cs [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.GenerateMember.GenerateParameterizedMember
+{
+    internal partial class AbstractGenerateParameterizedMemberService<TService, TSimpleNameSyntax, TExpressionSyntax, TInvocationExpressionSyntax>
+    {
+        private static async ValueTask<ITypeSymbol> ReplaceTypeParametersBasedOnTypeConstraintsAsync(
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/GenerateMember/GenerateParameterizedMember/TypeParameterSubstitution.cs:19]: thisis an extraction of an existing helper that has been moved here to make things simpler.

File: src/VisualStudio/Core/Def/Implementation/Progression/GraphQueries/SearchGraphQuery.cs [hunk 2]
```diff
@@ -50,25 +50,30 @@ private async Task ProcessProjectAsync(Project project, GraphBuilder graphBuilde
...
-                            await AddLinkedNodeForTypeAsync(project, namedType, graphBuilder, symbol.DeclaringSyntaxReferences.Select(d => d.SyntaxTree)).ConfigureAwait(false);
+                            await AddLinkedNodeForTypeAsync(
+                                project, symbol.WithSymbol(namedType), graphBuilder,
+                                symbol.Symbol.DeclaringSyntaxReferences.Select(d => d.SyntaxTree)).ConfigureAwait(false);
                         }
                         else
                         {
-                            await AddLinkedNodeForMemberAsync(project, symbol, graphBuilder).ConfigureAwait(false);
+                            await AddLinkedNodeForMemberAsync(
+                                project, symbol, graphBuilder).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/Progression/GraphQueries/SearchGraphQuery.cs:62]:
```suggestion
                            await AddLinkedNodeForMemberAsync(project, symbol, graphBuilder).ConfigureAwait(false);
```

File: src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentTypeFinder.cs [hunk 3]
```diff
@@ -151,16 +150,15 @@ internal static partial class DependentTypeFinder
...
             IImmutableSet<Project> projects,
             CancellationToken cancellationToken)
         {
             return FindTypesFromCacheOrComputeAsync(
                 type, solution, projects, s_typeToTransitivelyDerivedClassesMap,
-                c => FindDerivedClassesAsync(
-                    SymbolAndProjectId.Create(type, projectId: null), solution, projects,
-                    transitive: true, cancellationToken: c),
+                c => FindDerivedClassesAsync(type, solution, projects,
+                    transitive: true, c),
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentTypeFinder.cs:161]:
```suggestion
                c => FindDerivedClassesAsync(type, solution, projects, transitive: true, c),
```

File: src/Features/Core/Portable/GenerateMember/GenerateParameterizedMember/TypeParameterSubstitution.cs [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+            CancellationToken cancellationToken)
+        {
+            var visitor = new DetermineSubstitutionsVisitor(
+                compilation, availableTypeParameterNames, project, cancellationToken);
+
+            await visitor.Visit(type).ConfigureAwait(false);
+            return type.SubstituteTypes(visitor.Substitutions, compilation);
+        }
+
+        private class DetermineSubstitutionsVisitor : AsyncSymbolVisitor
```
Comment (mavasani) [inline src/Features/Core/Portable/GenerateMember/GenerateParameterizedMember/TypeParameterSubstitution.cs:33]: Nit: sealed

File: src/Features/Core/Portable/GenerateMember/GenerateParameterizedMember/TypeParameterSubstitution.cs (hunk 1 above)
Comment (mavasani) [reply src/Features/Core/Portable/GenerateMember/GenerateParameterizedMember/TypeParameterSubstitution.cs:19]: Does this file need reviewing or is it primarily code move?

File: src/VisualStudio/Core/Def/Implementation/Progression/GraphBuilder.cs [hunk 5]
```diff
@@ -160,7 +161,7 @@ public Document GetContextDocument(GraphNode node)
             }
         }
 
-        public ISymbol GetSymbol(GraphNode node)
+        public SymbolAndProjectId GetSymbol(GraphNode node)
```
Comment (mavasani) [inline src/VisualStudio/Core/Def/Implementation/Progression/GraphBuilder.cs:164]:
```suggestion
        public SymbolAndProjectId GetSymbolAndProjectId(GraphNode node)
```

File: src/VisualStudio/Core/Def/Implementation/Progression/GraphBuilder.cs [hunk 6]
```diff
@@ -169,15 +170,17 @@ public ISymbol GetSymbol(GraphNode node)
             }
         }
 
-        public Task<GraphNode> AddNodeForSymbolAsync(ISymbol symbol, GraphNode relatedNode)
+        public Task<GraphNode> AddNodeForSymbolAsync(SymbolAndProjectId symbol, GraphNode relatedNode)
```
Comment (mavasani) [inline src/VisualStudio/Core/Def/Implementation/Progression/GraphBuilder.cs:173]:
```suggestion
        public Task<GraphNode> AddNodeAsync(SymbolAndProjectId symbol, GraphNode relatedNode)
```

File: src/VisualStudio/Core/Def/Implementation/Progression/GraphBuilder.cs [hunk 7]
```diff
@@ -169,15 +170,17 @@ public ISymbol GetSymbol(GraphNode node)
...
         {
             // The lack of a lock here is acceptable, since each of the functions lock, and GetContextProject/GetContextDocument
             // never change for the same input.
             return AddNodeForSymbolAsync(symbol, GetContextProject(relatedNode), GetContextDocument(relatedNode));
         }
 
-        public async Task<GraphNode> AddNodeForSymbolAsync(ISymbol symbol, Project contextProject, Document contextDocument)
+        public async Task<GraphNode> AddNodeForSymbolAsync(SymbolAndProjectId symbolAndProjectId, Project contextProject, Document contextDocument)
```
Comment (mavasani) [inline src/VisualStudio/Core/Def/Implementation/Progression/GraphBuilder.cs:180]:
```suggestion
        public async Task<GraphNode> AddNodeAsync(SymbolAndProjectId symbolAndProjectId, Project contextProject, Document contextDocument)
```

File: src/VisualStudio/Core/Def/Implementation/Progression/GraphQueries/CallsGraphQuery.cs [hunk 8]
```diff
@@ -26,7 +22,7 @@ public async Task<GraphBuilder> GetGraphAsync(Solution solution, IGraphContext c
             foreach (var node in context.InputNodes)
             {
                 var symbol = graphBuilder.GetSymbol(node);
```
Comment (mavasani) [inline src/VisualStudio/Core/Def/Implementation/Progression/GraphQueries/CallsGraphQuery.cs:24]:
```suggestion
                var symbolAndProjectId = graphBuilder.GetSymbolAndProjectId(node);
```

File: src/VisualStudio/Core/Def/Implementation/Progression/GraphQueries/ContainsChildrenGraphQuery.cs [hunk 9]
```diff
@@ -25,7 +25,7 @@ public async Task<GraphBuilder> GetGraphAsync(Solution solution, IGraphContext c
                 {
                     var symbol = graphBuilder.GetSymbol(node);
```
Comment (mavasani) [inline src/VisualStudio/Core/Def/Implementation/Progression/GraphQueries/ContainsChildrenGraphQuery.cs:26]:
```suggestion
                    var symbolAndProjectId = graphBuilder.GetSymbolAndProjectId(node);
```

File: src/VisualStudio/Core/Def/Implementation/Progression/GraphQueries/SearchGraphQuery.cs [hunk 10]
```diff
@@ -63,12 +65,14 @@ private async Task ProcessProjectAsync(Project project, GraphBuilder graphBuilde
             }
         }
 
-        private async Task<GraphNode> AddLinkedNodeForTypeAsync(Project project, INamedTypeSymbol namedType, GraphBuilder graphBuilder, IEnumerable<SyntaxTree> syntaxTrees)
+        private async Task<GraphNode> AddLinkedNodeForTypeAsync(
+            Project project, SymbolAndProjectId<INamedTypeSymbol> namedType, GraphBuilder graphBuilder, IEnumerable<SyntaxTree> syntaxTrees)
```
Comment (mavasani) [inline src/VisualStudio/Core/Def/Implementation/Progression/GraphQueries/SearchGraphQuery.cs:69]:
```suggestion
            Project project, SymbolAndProjectId<INamedTypeSymbol> namedTypeAndProjectId, GraphBuilder graphBuilder, IEnumerable<SyntaxTree> syntaxTrees)
```

**Comment by CyrusNajmabadi** [comment]
> Does this file need reviewing or is it primarily code move?

It's primarily a move, and some cleanup.  But core logic is the same.

**Comment by mavasani** [comment]
> It's primarily a move, and some cleanup. But core logic is the same.

Thanks!

**Pull request merged**